File: src/pose_bridge.py
#!/usr/bin/env python3

import json
import subprocess

import rclpy
from rclpy.node import Node

from override_sim.msg import FieldElement, FieldElementArray


class PoseBridge(Node):
    """Polls Gazebo's dynamic pose topic and republishes the scoring objects
    (pins and cups) as a FieldElementArray on /_object_locations."""

    # color codes: 1 = red, 2 = blue, 3 = yellow
    PIN_COMBOS = {
        'ry': (1, 3),   # bottom red, top yellow
        'by': (2, 3),   # bottom blue, top yellow
        'yy': (3, 3),
        'rb': (1, 2),   # bottom red, top blue
    }

    NON_ELEMENT_NAMES = [
        'link', 'Otto', 'opponent', 'wheel', 'field', 'goal', 'toggle',
        'loader', 'ground',
    ]

    def __init__(self):
        super().__init__('pose_bridge')

        self.declare_parameter('world_name', 'override')
        self.world_name = self.get_parameter('world_name').value

        self.auto_update = self.create_timer(0.5, self.update_locations)
        self.element_locations = self.create_publisher(
            FieldElementArray, '/_object_locations', 10)

    def echo_gz_topic(self):
        """Fetch one JSON sample from the gz dynamic pose topic."""
        try:
            result = subprocess.run(
                ["gz", "topic", "-e", "-t",
                 f"world/{self.world_name}/dynamic_pose/info",
                 "-n", "1", "--json-output"],
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                text=True,
                timeout=5,
            )
        except subprocess.TimeoutExpired:
            self.get_logger().warning("gz topic command timed out after 5 seconds")
            return {"pose": []}
        except Exception as e:  # noqa: BLE001
            self.get_logger().error(f"Unexpected error running gz topic: {e}")
            return {"pose": []}

        if result.returncode != 0:
            self.get_logger().warning(
                f"gz topic returned code {result.returncode}: {result.stderr.strip()}")
            return {"pose": []}

        try:
            for line in result.stdout.strip().split('\n'):
                if line.strip():
                    try:
                        return json.loads(line)
                    except json.JSONDecodeError:
                        continue
            self.get_logger().warning("No valid JSON found in output")
            return {"pose": []}
        except Exception as e:  # noqa: BLE001
            self.get_logger().error(
                f"Unexpected error: {e}, stdout: '{result.stdout}', "
                f"stderr: '{result.stderr}'")
            return {"pose": []}

    @staticmethod
    def parse_pin_colors(name):
        """pin_ry_3 -> (bottom=1, top=3)"""
        for combo, (bottom, top) in PoseBridge.PIN_COMBOS.items():
            if f'pin_{combo}_' in name or name.startswith(f'pin_{combo}'):
                return bottom, top
        return 0, 0

    def update_locations(self):
        objects = FieldElementArray()
        data = self.echo_gz_topic()

        for pose in data.get("pose", []):
            name = pose.get("name", "")
            if any(tok in name for tok in self.NON_ELEMENT_NAMES):
                continue

            position = pose.get("position", {})
            if "x" not in position or "y" not in position or "z" not in position:
                continue

            el = FieldElement()
            el.object_name = name
            el.id = pose.get("id", 0)
            el.location.x = float(position["x"])
            el.location.y = float(position["y"])
            el.location.z = float(position["z"])
            o = pose.get("orientation", {})
            el.orientation.x = float(o.get("x", 0.0))
            el.orientation.y = float(o.get("y", 0.0))
            el.orientation.z = float(o.get("z", 0.0))
            el.orientation.w = float(o.get("w", 1.0))

            if 'cup' in name:
                el.element_type = 2
            elif 'pin' in name:
                el.element_type = 1
                el.bottom_color, el.top_color = self.parse_pin_colors(name)
            else:
                continue

            objects.elements.append(el)

        self.element_locations.publish(objects)


def main(args=None):
    rclpy.init(args=args)
    node = PoseBridge()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/pose_bridge.cpp
// pose_bridge.cpp - republishes the Gazebo dynamic-pose stream (pins and cups)
// as an override_sim/FieldElementArray on /_object_locations.
//
// The Gazebo topic /world/<world>/dynamic_pose/info is bridged to ROS by
// ros_gz_bridge (see sim_backend.launch.py) as ros_gz_interfaces/msg/Pose_V,
// which keeps each entity's name and id.
#include <memory>
#include <string>
#include <utility>
#include <vector>

#include <rclcpp/rclcpp.hpp>
#include <ros_gz_interfaces/msg/pose_v.hpp>

#include "override_sim/msg/field_element.hpp"
#include "override_sim/msg/field_element_array.hpp"

namespace override_sim
{

class PoseBridge : public rclcpp::Node
{
public:
  PoseBridge() : Node("pose_bridge")
  {
    this->declare_parameter<std::string>("world_name", "override");
    this->get_parameter("world_name", world_name_);

    pub_ = this->create_publisher<FieldElementArray>("/_object_locations", 10);
    sub_ = this->create_subscription<ros_gz_interfaces::msg::Pose_V>(
      "/world/" + world_name_ + "/dynamic_pose/info", rclcpp::SensorDataQoS(),
      [this](const ros_gz_interfaces::msg::Pose_V::SharedPtr msg) { latest_ = msg; });
    timer_ = this->create_wall_timer(
      std::chrono::milliseconds(500), [this]() { update_locations(); });
  }

private:
  // color codes: 1 = red, 2 = blue, 3 = yellow
  static std::pair<int, int> ParsePinColors(const std::string & name)
  {
    // pin_ry_3 -> (bottom=1, top=3)
    const std::vector<std::pair<std::string, std::pair<int, int>>> combos = {
      {"ry", {1, 3}}, {"by", {2, 3}}, {"yy", {3, 3}}, {"rb", {1, 2}}};
    for (const auto & c : combos) {
      const std::string token = "pin_" + c.first + "_";
      if (name.find(token) != std::string::npos ||
        name.rfind("pin_" + c.first, 0) == 0)
      {
        return c.second;
      }
    }
    return {0, 0};
  }

  static bool IsNonElement(const std::string & name)
  {
    static const std::vector<std::string> tokens = {
      "link", "Otto", "opponent", "wheel", "field", "goal", "toggle",
      "loader", "ground"};
    for (const auto & t : tokens) {
      if (name.find(t) != std::string::npos) {
        return true;
      }
    }
    return false;
  }

  void update_locations()
  {
    auto objects = std::make_unique<FieldElementArray>();
    if (latest_) {
      for (const auto & p : latest_->data) {
        if (IsNonElement(p.name)) {
          continue;
        }
        FieldElement el;
        el.object_name = p.name;
        el.id = static_cast<int32_t>(p.id);
        el.location = p.position;
        el.orientation = p.orientation;
        if (p.name.find("cup") != std::string::npos) {
          el.element_type = 2;
        } else if (p.name.find("pin") != std::string::npos) {
          el.element_type = 1;
          const auto colors = ParsePinColors(p.name);
          el.bottom_color = static_cast<int8_t>(colors.first);
          el.top_color = static_cast<int8_t>(colors.second);
        } else {
          continue;
        }
        objects->elements.push_back(el);
      }
    }
    pub_->publish(std::move(objects));
  }

  std::string world_name_;
  rclcpp::Publisher<FieldElementArray>::SharedPtr pub_;
  rclcpp::Subscription<ros_gz_interfaces::msg::Pose_V>::SharedPtr sub_;
  rclcpp::TimerBase::SharedPtr timer_;
  ros_gz_interfaces::msg::Pose_V::SharedPtr latest_;
};

}  // namespace override_sim

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<override_sim::PoseBridge>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
